File: Algorithms/HackerrankQuestions/Search/TripleSum.js
//https://www.hackerrank.com/challenges/triple-sum/problem?h_l=interview&playlist_slugs%5B%5D=interview-preparation-kit&playlist_slugs%5B%5D=search
function triplets(a, b, c) {
    const $ = arr => [...new Set(arr)].sort((a, b) => a - b);
    (a = $(a)), (b = $(b)), (c = $(c));
    // var is function scope, let is block scope, so put let below to declare count will not work
    for (var i = 0, j = 0, k = 0, count = 0; j < b.length; j++) {
        while (a[i] <= b[j]) i++;
        while (c[k] <= b[j]) k++;
        count += i * k;
    }
    return count;
}
function triplets2(a, b, c) {
    // sort a,b,c
    a.sort((a, b) => a - b);
    b.sort((a, b) => a - b);
    c.sort((a, b) => a - b);
    // get distinct a,b,c
    a = [...new Set(a)];
    b = [...new Set(b)];
    c = [...new Set(c)];
    let indexA = 0,
        indexB = 0,
        indexC = 0;
    let count = 0;
    while (indexB < b.length) {
        while (indexA < a.length && a[indexA] <= b[indexB]) {
            indexA++;
        }
        while (indexC < c.length && c[indexC] <= b[indexB]) {
            indexC++;
        }
        count += indexA * indexC;
        indexB++;
    }
    return count;
}

console.log(triplets([1, 3, 5], [2, 3], [1, 2, 3]));
console.log(triplets([1, 3, 5, 7], [5, 7, 9], [7, 9, 11, 13]));

File: Algorithms/HackerrankQuestions/Search/TripleSum.java
//https://www.hackerrank.com/challenges/triple-sum/problem?h_l=interview&playlist_slugs%5B%5D=interview-preparation-kit&playlist_slugs%5B%5D=search
package Algorithms.HackerrankQuestions.Search;

import java.util.*;

public class TripleSum {
  // Solution:
  static long triplets(int[] a, int[] b, int[] c) {
        Arrays.sort(a);
        a=Arrays.stream(a).distinct().toArray();
        Arrays.sort(b);
        b=Arrays.stream(b).distinct().toArray();
        Arrays.sort(c);
        c=Arrays.stream(c).distinct().toArray();
        int indexA=0,indexB=0,indexC=0;
        long count=0;
        while(indexB<b.length) {
            while(indexA<a.length && a[indexA] <= b[indexB])
                indexA++;
            while(indexC<c.length && c[indexC] <= b[indexB])
                indexC++;
            // convert indexes to long to prevent runtime error
            count+=(long)indexA*(long)indexC;
            indexB++;
        }
        return count;
    }
    public static void main(String[] args) {
        int[] a = new int[]{3,5,7};
        int b[] = new int[]{3,6};
        int[] c = new int[]{4,6,9};
        System.out.println(triplets(a,b,c));
         a = new int[]{1,4,5};
         b = new int[]{2,3,3};
         c = new int[]{1,2,3};
        System.out.println(triplets(a,b,c));
    }
}
